### ISSUE TITLE
Alacritty is slower than Terminator

### ISSUE THREAD

**Issue by manishrjain** [jwilm/alacritty#225]
I followed lead from another Github Issue here (against urxvt). Created a 1m.rdf file from github.com/dgraph-io/benchmarks/data/22million.rdf.gz using `zcat 22million.rdf.gz | head -n 1000000 > 1m.rdf`.

I then ran Alacritty and Terminator one after another, twice. Here're the results.

- Alacritty:  cat 1m.rdf  2.08s system 7% cpu 29.3 total
- Terminator: cat 1m.rdf  2.15s system 27% cpu 7.819 total

- Alacritty: cat 1m.rdf 2.34s system 6% cpu 37.762 total
- Terminator: cat 1m.rdf  1.98s system 25% cpu 7.680 total

Alacritty is a factor slower than Terminator. Here's what archey3 shows on my system.

```
               +                OS: Arch Linux x86_64
               #                Hostname: mrjn-thinkpad
              ###               Kernel Release: 4.8.13-1-ARCH
             #####              Uptime: 1:04
             ######             WM: i3
            ; #####;            DE: None
           +##.#####            Packages: 764
          +##########           RAM: 2498 MB / 7855 MB
         #############;         Processor Type: Intel(R) Core(TM) i5-6300U CPU @ 2.40GHz
        ###############+        $EDITOR: vim
       #######   #######        Root: 69G / 168G (41%) (ext4)
     .######;     ;###;`".      
    .#######;     ;#####.       
    #########.   .########`     
   ######'           '######    
  ;####                 ####;   
  ##'                     '##   
 #'                         `#  
```

**Comment by jwilm** [comment]
Are you using the Mesa driver by chance?

**Comment by manishrjain** [comment]
This is the relevant output from `lspci -v`

```
00:02.0 VGA compatible controller: Intel Corporation HD Graphics 520 (rev 07) (prog-if 00 [VGA controller])
	Subsystem: Lenovo Device 5053
	Flags: bus master, fast devsel, latency 0, IRQ 125
	Memory at f0000000 (64-bit, non-prefetchable) [size=16M]
	Memory at e0000000 (64-bit, prefetchable) [size=256M]
	I/O ports at e000 [size=64]
	[virtual] Expansion ROM at 000c0000 [disabled] [size=128K]
	Capabilities: <access denied>
	Kernel driver in use: i915
	Kernel modules: i915
```

**Comment by jwilm** [comment]
Could you post the output of `glxinfo | grep OpenGL` please?

**Comment by mbilker** [comment]
Intel graphics on Linux goes through Mesa.

**Comment by manishrjain** [comment]
I don't have glxinfo. Seems like you need to install Mesa to provide that.

This is from `/etc/X11/xorg.conf.d/20-intel.conf`:

```
Section "Device"
  Identifier  "Intel Graphics"
  Driver      "intel"
  #Option      "DRI" "2"             # DRI3 is now default 
  #Option      "AccelMethod"  "sna" # default
  #Option      "AccelMethod"  "uxa" # fallback
EndSection
```

**Comment by fd0** [comment]
Same issue here, alacritty is way slower than my default terminal (`rxvt-unicode`): `dmesg` takes 0.035s vs 1.75s in alacritty. It seems I'm on the same hardware and software platform as @manishrjain (Arch Linux on Intel Skylake):
```
OpenGL vendor string: Intel Open Source Technology Center
OpenGL renderer string: Mesa DRI Intel(R) HD Graphics 530 (Skylake GT2)
OpenGL core profile version string: 4.5 (Core Profile) Mesa 13.0.3
OpenGL core profile shading language version string: 4.50
OpenGL core profile context flags: (none)
OpenGL core profile profile mask: core profile
OpenGL core profile extensions:
OpenGL version string: 3.0 Mesa 13.0.3
OpenGL shading language version string: 1.30
OpenGL context flags: (none)
OpenGL extensions:
OpenGL ES profile version string: OpenGL ES 3.2 Mesa 13.0.3
OpenGL ES profile shading language version string: OpenGL ES GLSL ES 3.20
OpenGL ES profile extensions:
```

**Comment by lordtoniok** [comment]
Using mesa git with r600 i have a huge difference 2.53s with st to cat 1m.rdf and 1m40s for alacritty.

**Comment by xeyownt** [comment]
For info, I experience the same issue. Also Intel HW, but older. This is on Ubuntu 16.04. With `cat`, the scrolling is much smoother on alacritty but also much slower to print the whole file. But with `vim`/`nvim` (without tmux), the output is both slower and less smooth (can see partial page refresh for instance).

```
OpenGL vendor string: Intel Open Source Technology Center
OpenGL renderer string: Mesa DRI Intel(R) Haswell Mobile 
OpenGL core profile version string: 3.3 (Core Profile) Mesa 11.2.0
OpenGL core profile shading language version string: 3.30
OpenGL core profile context flags: (none)
OpenGL core profile profile mask: core profile
OpenGL core profile extensions:
OpenGL version string: 3.0 Mesa 11.2.0
OpenGL shading language version string: 1.30
OpenGL context flags: (none)
OpenGL extensions:
OpenGL ES profile version string: OpenGL ES 3.0 Mesa 11.2.0
OpenGL ES profile shading language version string: OpenGL ES GLSL ES 3.00
OpenGL ES profile extensions:
```

**Comment by xeyownt** [comment]
Looking other bugs, this seems a duplicate of #125.

**Comment by jwilm** [comment]
@xeyownt there's a thing in the FAQ about this, but it's probably not very visible. Maybe we should make an ISSUES.md highlighting a few common known issues.

> _It's not fast! Why?_ There's a known bug affecting some versions of Mesa/libxcb where calls to glClear take an insanely long time. If it's not that, there's probably another bug. I'd be happy to look at the issue if you can provide some profiling information (wall time and otherwise).

**Comment by xeyownt** [comment]
Ah ah, yes. I read the FAQ few hours before posting but completely forgot that one indeed. Maybe I missed the intel keyword, or was still in the mood "this bug cannot affect me".

Is the bug really related to slow glClear, or is due to synchronizing queued work? 

For reading a bit on the subject, it seems that we can discriminate by adding a call to glFinish right before calling glClear (see for instance [SO 29551516](http://stackoverflow.com/questions/29551516/glclear-takes-too-long-android-opengl-es-2) and [SO 12224230](http://stackoverflow.com/questions/12224230/why-is-glclear-blocking-in-opengles)). [OpenGL Common Mistakes](https://www.khronos.org/opengl/wiki/Common_Mistakes) is also quite informative. Note that I'm not an OpenGL expert so please excuse me if I say obvious things.

**Comment by jwilm** [comment]
Should be resolved as of #410

**Comment by xeyownt** [comment]
I confirm issue is resolved for me. Now Alacritty is twice faster than Terminator when `cat`-ing a 7.5MB file. Scrolling in vim/nvim is a bit faster, no more jerkiness.
